include/deal.II/lac/sparse_matrix.templates.h

```diff
@@ -1898,6 +1898,80 @@ SparseMatrix<number>::get_sparsity_pattern () const
 }
 
 
+namespace internal
+{
+  namespace SparseMatrix
+  {
+    template <class StreamType, typename realType>
+    inline
+    void print_value(StreamType &out, const realType&val)
+    {
+      out << val;
+    }
+
+    template<class StreamType, typename realType>
+    inline
+    void print_value(StreamType &out, const std::complex<realType> &val)
+    {
+      // the following format is what is required by the matrix market format.
+      out << val.real << ' ' << val.imag;
+    }
+
+    template<class StreamType, class numberType>
+    void
+    print_header(StreamType &out,const numberType & dummy)
+    {
```

Comment:
1. Good question. I don't know. It's a pretty commonly used way to dispatch a function call to several others. This idiom of partially specializating a class (or struct) template is commonly used because C++ does not allow for partial specialization of function templates (which is, really, what you want to do here).

2. A `static` member function can be used without an object. I.e., you can call it as in `PrintHeader<...>::print_header(...)`, rather than having to create an object as in
```
  PrintHeader<....> tmp;   // object of this type
  tmp.print_header (out);  // use a regular member function of the object
```
This is possible here because the structure has no member variables, and so there is no state that the member function could access anyway.

3/ Yes, of course.